neuralmonkey/decoding_function.py

```diff
@@ -1,59 +1,93 @@
-                                       [-1, state_size])
-
-        state = initial_state
-        outputs = []
-        prev = None
-
-        attns = [a.initialize(batch_size, dtype) for a in attention_objects]
-
-        states = []
-        for i, inp in enumerate(decoder_inputs):
-            if i > 0:
-                tf.get_variable_scope().reuse_variables()
-            # If loop_function is set, we use it instead of decoder_inputs.
-            if loop_function is not None and prev is not None:
-                with tf.variable_scope("loop_function", reuse=True):
-                    inp = loop_function(prev, i)
-            # Merge input and previous attentions into one vector of the right
-            # size.
-            x = tf.nn.seq2seq.linear([inp] + attns, embedding_size, True)
-            # Run the RNN.
-            cell_output, state = cell(x, state)
-            states.append(state)
-            # Run the attention mechanism.
-            attns = [a.attention(state) for a in attention_objects]
 
-            if attns:
-                with tf.variable_scope("AttnOutputProjection"):
-                    output = tf.nn.seq2seq.linear([cell_output] + attns, output_size,
-                                             True)
+        output, state = decode_step(decoder_inputs[0], initial_state,
+                                    attention_objects, cell,
+                                    maxout_size, output_size)
+        outputs.append(output)
+        states.append(state)
+
+        for step in range(1, len(decoder_inputs)):
+            tf.get_variable_scope().reuse_variables()
+
+            if loop_function:
+                decoder_input = loop_function(state, step)
             else:
-                output = cell_output
+                decoder_input = decoder_inputs[step]
 
-            if loop_function is not None:
-                prev = output
+            output, state = decode_step(decoder_input, state, attention_objects,
+                                        cell, maxout_size, output_size)
             outputs.append(output)
+            states.append(state)
 
     return outputs, states
 
+
+def decode_step(prev_output, prev_state, attention_objects, rnn_cell,
+                maxout_size, output_size):
+    ## compute c_i:
+    contexts = [a.attention(prev_state) for a in attention_objects]
+
+    ## compute prob(y_i | s_i-1, y_i-1, c_i)
+    output = maxout_projection(prev_state, prev_output, contexts,
+                               maxout_size, output_size)
+
+    ## compute s_i based on y_i-1, c_i and s_i-1
+    _, state = rnn_cell(tf.concat(1, [prev_output] + contexts), prev_state)
+
+    return output, state
+
+
+def linear_projection(states, size):
+    with tf.variable_scope("AttnOutputProjection"):
+        output = tf.nn.seq2seq.linear(states, size, True)
+        return output
+
+
+def maxout_projection(prev_state, prev_output, current_contexts,
```

Comment:
Nemělo by se to odstěhovat někam do `nn` jako maxout_layer? 

Úplně nejhezčí a nejobecnější by asi bylo, kdyby ta projekce byla instance multilayer-perceptronu, kterej by se ale musel zobecnit, aby:
* na konci nemusel mít softmax
* mohl mít kromě dense layers ještě pofiderní vrstvy jako je highway layer a max-out

To asi není důležitý, tak jestli se ti s tím nechce teď dělat, tak se to asi může napsat do nějaké issue do budoucna.